apis/memorystore/v1alpha1/instance_types.go

```diff
@@ -0,0 +1,198 @@
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/apis/k8s/v1alpha1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var MemorystoreInstanceGVK = GroupVersion.WithKind("MemorystoreInstance")
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// MemorystoreInstanceSpec defines the desired state of MemorystoreInstance
+// +kcc:proto=google.cloud.memorystore.v1beta.Instance
+type MemorystoreInstanceSpec struct {
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="ResourceID field is immutable"
+	// Immutable.
+	// The MemorystoreInstance name. If not given, the metadata.name will be used.
+	ResourceID *string `json:"resourceID,omitempty"`
+
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Parent field is immutable"
+	// Immutable.
+	// +required
+	Parent `json:",inline"`
+
+	// Optional. Labels to represent user-provided metadata.
+	Labels map[string]string `json:"labels,omitempty"`
+
+	// Optional. Number of replica nodes per shard. If omitted the default is 0
+	//  replicas.
+	ReplicaCount *int32 `json:"replicaCount,omitempty"`
+
+	// Optional. Immutable. Authorization mode of the instance.
+	AuthorizationMode *string `json:"authorizationMode,omitempty"`
+
+	// Optional. Immutable. In-transit encryption mode of the instance.
+	TransitEncryptionMode *string `json:"transitEncryptionMode,omitempty"`
+
+	// Optional. Number of shards for the instance.
+	ShardCount *int32 `fimmujson:"shardCount,omitempty"`
+
+	// Optional. Immutable. Machine type for individual nodes of the instance.
+	NodeType *string `json:"nodeType,omitempty"`
+
+	// Optional. Persistence configuration of the instance.
+	PersistenceConfig *PersistenceConfig `json:"persistenceConfig,omitempty"`
+
+	// Optional. Immutable. Engine version of the instance.
+	EngineVersion *string `json:"engineVersion,omitempty"`
+
+	// Optional. User-provided engine configurations for the instance.
+	EngineConfigs map[string]string `json:"engineConfigs,omitempty"`
+
+	// Optional. Immutable. Zone distribution configuration of the instance for
+	//  node allocatiteon.
+	ZoneDistributionConfig *ZoneDistributionConfig `json:"zoneDistributionConfig,omitempty"`
+
+	// Optional. If set to true deletion of the instance will fail.
+	DeletionProtectionEnabled *bool `json:"deletionProtectionEnabled,omitempty"`
+
+	// Required. Immutable. User inputs and resource details of the auto-created
+	//  PSC connections.
+	PscAutoConnections []PscAutoConnection `json:"pscAutoConnections,omitempty"`
+
+	// Optional. Endpoints for the instance.
+	Endpoints []Instance_InstanceEndpoint `json:"endpoints,omitempty"`
+
+	// Optional. The mode config for the instance.
+	Mode *string `json:"mode,omitempty"`
+}
+
+type Parent struct {
+	// +required
+	ProjectRef *refs.ProjectRef `json:"projectRef"`
+
+	// +required
+	Location string `json:"location"`
+}
```

Comment:
We do not need the parent. We can keep ProjectRef and Location as part of Spec directly